### ISSUE TITLE
Service's IP address leak, even if it is a hidden service

### ISSUE THREAD

**Issue by ghost** [searx/searx#2197]
You can use morty proxy to proxy those "what's my IP address" websites, being a proxy, it will use the server's IP address to query the website. Even if it is a hidden service. Leaking it
![1600119909](https://user-images.githubusercontent.com/53711069/93142079-9c648b00-f6e5-11ea-8311-792b079e27da.png)
(This was made in I2P)
Also, if you search "What's my IP" it will also leak the IP address
![1600119789](https://user-images.githubusercontent.com/53711069/93142105-a4bcc600-f6e5-11ea-8faa-c1bdae2592a7.png)

This can be also done in .onion services
![1600120014](https://user-images.githubusercontent.com/53711069/93142159-be5e0d80-f6e5-11ea-8a94-6d3d5b68475a.png)

**Possible solutions:** 
* Encourage hidden services searx instances' owners to disable duckduckgo definitions in their searx installation
* Discourage hidden services searx instances' owners to use Morty Proxy

**Comment by unixfox** [comment]
I can't reproduce the first bug, when I search for "What's my IP" on my instance: https://searx.be/?q=what%27s+my+IP+address& and then use the "proxied" link on the `https://www.whatismyip.com` website I don't get my personal IP:
![image](https://user-images.githubusercontent.com/4016501/93243153-78aa4f00-f790-11ea-987c-1afc204fbc4f.png)

And for the ddg definitions, I'm unable to use this engine because it's broken (I tested it on multiple instances):
![image](https://user-images.githubusercontent.com/4016501/93243628-23227200-f791-11ea-95e6-19c11286f25d.png)

**Comment by ghost** [comment]
@unixfox 
Weird, well, ddg definitions worked for me.
![1600191337](https://user-images.githubusercontent.com/53711069/93244890-b3a98400-f78a-11ea-8a9a-056f53cdd57a.png)

Also, your proxied IP address and mine are the same
![1600191280](https://user-images.githubusercontent.com/53711069/93244876-adb3a300-f78a-11ea-8c8b-dca806e38556.png)

Probably it gives a different IP address because you're using Morty Proxy in a different server.

**Comment by unixfox** [comment]
Sorry I read too quickly your initial question, I thought your personal IP address was leaking because you tried to hide the IP address on the first screenshot. My bad.

----------------
Is this really an issue that the IP address of the server that is hosting the hidden service is leaking? I thought the point of using a hidden service is just for enhance the anonymity of the user by avoiding sending personal information like the IP address and so on.

**Comment by dalf** [comment]
Side note: it is difficult to hide the server IP.

Basically:
* it requires the searx instance to know the public IP. An instance can listen on private IP, so it is not easy to get the public IP list without manual entry and update.
* search for the IPs in the results / answer / infoboxes and remove them.

**Comment by ghost** [comment]
>Is this really an issue that the IP address of the server that is hosting the hidden service is leaking? I thought the point of using a hidden service is just for enhance the anonymity of the user by avoiding sending personal information like the IP address and so on.

Yes. I know, but I think that the purpose of a hidden service is a) hide the server IP address and b) hide the user's IP address. Specially the first one.

>search for the IPs in the results / answer / infoboxes and remove them.

I think that could work  
But like the opening post, discouraging administrators for using morty proxy in hidden services could also be a good idea. Specially in non outproxy hidden networks (Like I2P)

**Comment by unixfox** [comment]
> > Is this really an issue that the IP address of the server that is hosting the hidden service is leaking? I thought the point of using a hidden service is just for enhance the anonymity of the user by avoiding sending personal information like the IP address and so on.
> 
> Yes. I know, but I think that the purpose of a hidden service is a) hide the server IP address and b) hide the user's IP address. Specially the first one.

It's important for the services that doesn't want to have their server down by denial of services attacks but for Searx I don't see the point of attacking a specific instance when there are multiple other ones.

Apart from this reason I don't understand why a Searx instance running on a hidden service would want to hide its real server IP address.

@qorg11 Do you have other point/argument against the leak of the IP address of the searx instance on a hidden service?

> Side note: it is difficult to hide the server IP.
> 
> Basically:
> 
>     * it requires the searx instance to know the public IP. An instance can listen on private IP, so it is not easy to get the public IP list without manual entry and update.
> 
>     * search for the IPs in the results / answer / infoboxes and remove them.

That's too tricky to do that I don't think it's worth it. It's like playing cat and mouse but with some text that could contain the public IP address of the instance.

**Comment by dalf** [comment]
> That's too tricky to do that I don't think it's worth it. It's like playing cat and mouse but with some text that could contain the public IP address of the instance.

I agree and it gives a false sense of "it is working". Example: an engine returns the CIDR related to the server IP, or the geoip result of the server.

Nevertheless, the server answers is miss leading when the user search for:
* _ip_: the server and user IP.
* _what is my ip ?_ : the server IP.

Possible workaround: add the engine or plugin name for each answer.

**Comment by unixfox** [comment]
> > That's too tricky to do that I don't think it's worth it. It's like playing cat and mouse but with some text that could contain the public IP address of the instance.
> 
> I agree and it gives a false sense of "it is working". Example: an engine returns the CIDR related to the server IP, or the geoip result of the server.
> 
> Nevertheless, the server answers is miss leading when the user search for:
> * _ip_: the server and user IP.
> * _what is my ip ?_ : the server IP.
> 
> Possible workaround: add the engine or plugin name for each answer.

It won't work against someone that is dedicated to get the public IP address of the hidden service.

An example is someone index a website that he is control of. He just have to bypass the protection included into Searx/morty and bingo he got the IP address.

**Comment by mohe2015** [comment]
The only correct method for hidden services would be to use tor to do the search. Which probably doesn't work well as tor is usually blocked. So I agree that there is nothing you can do about that. Disabling morty proxy may be a useful recommendation but as already said e.g. duckduckgo can still leak the ip and also the results may leak the geolocation. You could also use a proxy which would then leak the proxy ip and not that of the hidden service.

**Comment by dalf** [comment]
@unixfox : "_Possible workaround_" is misleading : it is not a workaround about the issue, but the UX when the user search for "ip" or "what is my ip". The user still see two IPs : one from the plugin "self_info", one from an engine. But at least the user can spot which IP is the right one (the one from the self_info plugin).

@mohe2015 Tor won't hide the IP as @unixfox said in the comment just above.

**Comment by mohe2015** [comment]
@dalf It should if tor is used for the search request from searx to the search provider too.

**Comment by MarcAbonce** [comment]
I haven't tried it with morty, but in searx if you set the proxies in `settings.yml` to a Tor socket the engines shouldn't be able to tell where is your hidden service.

Here's an example I have that uses Tor to connect to the engines: http://7ea7vegvcqgdzxjz.onion based on the branch in https://github.com/searx/searx/pull/565. But again, it doesn't have morty so this doesn't really answer the question directly.

So yeah, as others have already mentioned, searx is only meant to hide its users but not the server itself. However, it is possible, at least in principle, to hide the server too by proxying both searx and morty's outgoing connections through Tor.

**Comment by resynth1943** [comment]
> search for the IPs in the results / answer / infoboxes and remove them.

I don't understand. Wouldn't the results for "What's my IP?" just leak the engines' IP addresses?

**Comment by mohe2015** [comment]
@resynth1943 normally yes but e.g. duckduckgo allows you to search for "!ip" and then the search engines shows you the ip of the search requester which is the one from searx. Also proxied results from morty would show searx ip too.
